### PR TITLE
Review request for Identity to add version stable/2022-06-01

### LINKED ISSUE #18311: [Communication Services - AUTH] API Stewardship Review
## Complete Azure API review for GA for CTE flows

**Service name:** ACS Auth

Key contact for this review: Petr Svihlik

**PR Link:** https://github.com/Azure/azure-rest-api-specs/pull/17930

**Champion scenarios** that briefly explain how developers will use your API:

- The main goal of this feature is to allow developers who use ACS (directly or via SDKs) to build and integrate a custom user interface for calling functionality with Microsoft 365 Teams identities.
- A typical scenario is building an attendant console (e.g. for receptionists to be able to accept incoming calls and redirect them to the right person). 
- Developers will use the new API to exchange AAD Access tokens (of AAD users with a Teams license) for Skypetokens. These Skypetokens can then be used to call ACS APIs directly or via SDKs.
- Changes for GA:
- Unifying the endpoint name with the method name in the SDKs (As this is a breaking change compared to the Public Preview, we’ll need help determining whether we need to create a new breaking change intake and potentially some guidance.)
- Adding two more parameters to allow Contoso restrict who can exchange the token.

### PR DESCRIPTION
**Pull request by petrsvihlik**
## Champion scenario
Changes in this PR relate to a feature known as [Custom Teams Endpoint](https://docs.microsoft.com/en-us/azure/communication-services/quickstarts/manage-teams-identity). The goal of the feature is to allow developers to build custom calling experiences using ACS while utilizing (M365) Teams identities.​

## Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new API version **(described below in the "Champion scenario" section)**
  2. When are you targeting to deploy the new ~~service/~~ feature to public regions?
     - **End of June 2022**
  3. When do you expect to publish the swagger?
     - **March 2022**
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js

## Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

## Review links
For reviewers' convenience, we provide a couple of links
- GitHub [diff](https://github.com/petrsvihlik/CommunicationIdentity/compare/68533f967a41e6593ce3146a50abb2b8578cac41..00781c77990fb52dfe3a74ed7cb6d10cd5b2ec13) of Public Preview (`2021-10-31-preview`) vs GA (`2022-06-01`)
- [editor.swagger.io](https://editor.swagger.io/?url=https://raw.githubusercontent.com/petrsvihlik/azure-rest-api-specs/communication-Identity-2022-06-01/specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json)
- [OpenAPI Hub diff](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYWluL3NwZWNpZmljYXRpb24vY29tbXVuaWNhdGlvbi9kYXRhLXBsYW5lL0lkZW50aXR5L3JlYWRtZS5tZCIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9wZXRyc3ZpaGxpay9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iL2NvbW11bmljYXRpb24tSWRlbnRpdHktMjAyMi0wNi0wMS9zcGVjaWZpY2F0aW9uL2NvbW11bmljYXRpb24vZGF0YS1wbGFuZS9JZGVudGl0eS9yZWFkbWUubWQiLCJzb3VyY2VUYWciOiJwYWNrYWdlLXByZXZpZXctMjAyMS0xMCIsInRhcmdldFRhZyI6InBhY2thZ2UtMjAyMi0wNiJ9)
- [OpenAPI Hub validation](https://portal.azure-devex-tools.com/branch/petrsvihlik/azure-rest-api-specs/communication-Identity-2022-06-01...main/)

## Changes
There are two changes in the swagger compared to Public Preview:
1. `TeamsUserAccessTokenRequest` renamed to `TeamsUserExchangeTokenRequest`
2. `/teamsUser/:getToken` - on top of existing `token` parameter, we add two new parameters - `appId` and `userId` allowing Contoso to specify the AAD user and app that are allowed to perform the token exchange. 

### Problem definition
Contoso sends a PFT token obtained on the side of their user to ACS Auth. Because the PFT is encrypted, Contoso can’t verify its content and confirm that it belongs to their AAD app registration. It would be possible that a malicious actor could use an access token issued for a different 3P app or a different user which poses two main concerns:

- **Consent** - Fabrikam admin may give consent to an application made by Contoso. Without the validation, the AAD token issued for Contoso app could be used to mint skypetoken in an app different that the Contoso's one.
- **Billing** - Contoso is billed if a skypetoken minted for them is used in ACS. Without the validation, Contoso may think they pay for Alice who asked for the skypetoken, however they may unintentionally pay for Betty for whom the AAD token has been issued.
Custom Teams Endpoint - Possible malicious use schema

  - The related design can be found at [ACS-Auth-Verify-AppID-in-CTE.docx](https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/EVpb4TM-4-xKqgJVoLVDB9QBdianJEk8PGc70WrFqPKihg?e=JeCc73&wdOrigin=TEAMS-ELECTRON.p2p.bim&wdExp=TEAMS-TREATMENT&wdhostclicktime=1645351388526)

### Problem mitigation
Adding the following two parameters to the endpoint so that ACS Auth service can verify them:
- App Client ID - Contoso's backend app should generally be aware of this information.
- OID - the premise is that Contoso already has AuthNZ (+ an AAD app) in place to secure their backend. The expected flow is that the Contoso's backend will use an `oid` retrieved from the bearer token (both [access tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens) and [ID tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens) contain `oid`) sent as an authorization header
Custom Teams Endpoint - Malicious schema mitigation via verifying AppID + OID

## ~~Breaking Change Review Checklist~~ No breaking changes
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by petrsvihlik** [comment]
OpenAPI Hub: https://portal.azure-devex-tools.com/branch/petrsvihlik/azure-rest-api-specs/communication-Identity-2022-06-01...main/

File: specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+          }
+        },
+        "x-ms-examples": {
+          "Revoke access tokens": {
+            "$ref": "./examples/RevokeAccessTokens.json"
+          }
+        }
+      }
+    },
+    "/teamsUser/:getToken": {
```
Comment (anjulgarg) [inline specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json:136]: Reading this API path makes me think that this returns a teams user's access token.
My recommendation would be something like 
`/teamsUser/:getCommunicationToken`

File: specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+          "200": {
+            "description": "Success",
+            "schema": {
+              "$ref": "#/definitions/CommunicationIdentityAccessToken"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Exchange a Teams user access token": {
+            "$ref": "./examples/GetTeamsUserAccessToken.json"
```
Comment (anjulgarg) [inline specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json:179]: Example states that an `endpoint` attribute is required but it is not documented anywhere

File: specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json (hunk 1 above)
Comment (petrsvihlik) [reply specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json:136]: we'll revert back to `exchangeAccessToken`

File: specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json (hunk 2 above)
Comment (petrsvihlik) [reply specification/communication/data-plane/Identity/stable/2022-06-01/CommunicationIdentity.json:179]: it's the path of the request (discussed during the internal review)

**Comment by petrsvihlik** [comment]
sample app: https://github.com/petrsvihlik/azure-communication-identity-cte-samples

**Comment by petrsvihlik** [comment]
@markweitzel Can we get this PR merged please? ([sign-off here](https://github.com/Azure/azure-rest-api-specs/issues/18311#issuecomment-1076529014))

**Comment by petrsvihlik** [comment]
@jhendrixMSFT can you please help merging the PR?

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**